src/canvasObject.js

```diff
@@ -26,111 +44,88 @@ var CanvasObject = function(config, dispatcher) {
 CanvasObject.prototype = {
   openTileSource: function(viewer) {
     var self = this;
-
-    // We've already loaded this tilesource instead of the thumbnail
-    if(this.fullyOpened) {
-      return;
-    }
-
-    // otherwise, continue loading the tileSource.
-    this.dispatcher.emit('detail-tile-source-requested', { 'detail': this.id });
-    viewer.addTiledImage({
-      x: this.x,
-      y: this.y,
-      width: this.width,
-      tileSource: this.tileSourceUrl,
-      opacity: this.opacity,
-      clip: this.clipRegion,
-      index: 0, // Add the new image below the stand-in.
-
-      success: function(event) {
-        var main = event.item;
-        self.fullyOpened = true;
-
-        var tileDrawnHandler = function(event) {
-          if (event.tiledImage === main) {
-            var previousImageObj = self._mainImageObj;
-
-            viewer.removeHandler('tile-drawn', tileDrawnHandler);
-            main.setOpacity(0, true);
-            self._fade(main, 1);
-            self._setMainImage(main);
-
-            if(previousImageObj){
-              viewer.world.removeItem(previousImageObj);
-            }
-            self.dispatcher.emit('detail-tile-source-opened', { 'detail': self.id });
-          }
-        };
-        viewer.addHandler('tile-drawn', tileDrawnHandler);
-      },
-
-      error: function(event) {
-        var errorInfo = {
-          id: self.id,
-          message: event.message,
-          source: event.source
-        };
-        self.dispatcher.emit('detail-tile-source-failed', {'detail': errorInfo});
+    var onTileDrawn = function(event) {
+      var main = event.tiledImage;
+      main.setOpacity(0, true);
```

Comment:
Seems like instead of passing the OSD event straight through, you should pass the image resource, and things like setting the opacity and fading should happen through the image resource. That way there is good separation of concerns.